modules/hub/src/PaymentsService.ts

```diff
@@ -144,6 +168,75 @@ export default class PaymentsService {
     }
   }
 
+  // TODO: Need to check with Rahul about whether or not the public
+  // wrapper is needed and how to handle a meta object here (same as 
+  // above?)
+  public async doRedeem(
+    user: string,
+    secret: string,
+  ): Promise<MaybeResult<{ purchaseId: string }>> {
+    return this.db.withTransaction(() => this._doRedeem(user, secret))
+  }
+
+  private async _doRedeem(user: string, secret: string): Promise<MaybeResult<{ purchaseId: string | null }>> {
+    const channel = await this.channelsDao.getChannelOrInitialState(user)
+    // channel checks
+    if (channel.status !== 'CS_OPEN') {
+      throw new Error('Channel is not open: ' + user)
+    }
+
+    const payment = await this.paymentMetaDao.getLinkedPayment(secret)
+    // if hub can afford the payment, sign and forward payment
+    // otherwise, collateralize the channel
+    const prev = convertChannelState('bn', channel.state)
+    const amt = convertPayment('bn', payment.amount)
+    let purchasePayment = null
+    if (!this.validator.cantAffordFromBalance(prev, amt, "hub")) {
```

Comment:
This is a little confusing with the double negative. Why don't we do something like:
```
// doCollateralizeIfNecessary here
if (this.validator.cantAffordFromBalance(prev, amt, "hub")) {
  // return null here, or return an error
}

// rest of function assuming happy case
```

Reply:
Something Slipper told me one time: It's better and more readable to not nest the happy case stuff in `if` statements, use the `if`s to check invariants and return/error out of the function early, so the stuff outside of the if's is all the normal/happy case stuff.